{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "30cf9e59",
   "metadata": {},
   "source": [
    "## Use case 1: Obtain germline allele frequencies given gene and mutation entities\n",
    "- Test passing gene and mutation entities to the gnomad API to retrieve germline frequencies\n",
    "- documentation here https://gnomad.broadinstitute.org/data#api\n",
    "- needs gql, a python graphQL client installed (which depends on aiohttp)\n",
    "- pip install gql and aiohttp in your env "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "5b304f09",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.environ[\"CUDA_VISIBLE_DEVICES\"] = \"0\" # note the GPU index as a string "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "18df88d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "from pathlib import Path\n",
    "\n",
    "# this notebook is in notebooks — go up one level\n",
    "project_root = Path().resolve().parent\n",
    "sys.path.append(str(project_root))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "dca3997e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from methods.gdc_api_calls import get_ssm_id\n",
    "from gql import gql, Client\n",
    "from gql.transport.aiohttp import AIOHTTPTransport"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1e78c3ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aartiv/.pyenv/versions/3.9.2/lib/python3.9/site-packages/gql/transport/aiohttp.py:92: UserWarning: WARNING: By default, AIOHTTPTransport does not verify ssl certificates. This will be fixed in the next major version. You can set ssl=True to force the ssl certificate verification or ssl=False to disable this warning\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "transport = AIOHTTPTransport(url=\"https://gnomad.broadinstitute.org/api\")\n",
    "client = Client(transport=transport, fetch_schema_from_transport=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7220b625",
   "metadata": {},
   "outputs": [],
   "source": [
    "async def get_germline_af(input):\n",
    "  query = gql(\n",
    "    \"\"\"\n",
    "    query VariantsInGene($gene_symbol: String!) {\n",
    "      gene(gene_symbol: $gene_symbol, reference_genome: GRCh38) {\n",
    "        variants(dataset: gnomad_r4) {\n",
    "          variant_id\n",
    "          pos\n",
    "          exome {\n",
    "            ac\n",
    "            ac_hemi\n",
    "            ac_hom\n",
    "            an\n",
    "            af\n",
    "          }\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "    \"\"\"\n",
    "  )\n",
    "  result = await client.execute_async(query, variable_values=input)\n",
    "  return result\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "93264993",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_start_position(ssm_id):\n",
    "    ssms_endpt = 'https://api.gdc.cancer.gov/ssms/{}?fields=start_position'.format(ssm_id)\n",
    "    output = requests.get(ssms_endpt).json()\n",
    "    start_pos = output['data']['start_position']\n",
    "    return start_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8fabc777",
   "metadata": {},
   "outputs": [],
   "source": [
    "async def compute_gnomad_af(gene, mutation):\n",
    "    germline_af = {}\n",
    "    ssm_id = get_ssm_id(gene=gene, mutation=mutation)\n",
    "    start_position = get_start_position(ssm_id=ssm_id)\n",
    "    input = {\"gene_symbol\": gene }\n",
    "    result = await get_germline_af(input)\n",
    "    for d in result['gene']['variants']:\n",
    "        if d['pos'] == start_position:\n",
    "            germline_af[d['variant_id']] = d['exome']['af']\n",
    "    return germline_af"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d3b97b52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'12-25245350-C-G': 1.369332163010778e-06,\n",
       " '12-25245350-C-T': 2.7386718263243876e-06,\n",
       " '12-25245350-C-A': 6.84666081505389e-07}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# super low germline frequency as expected for the G12D variant\n",
    "await compute_gnomad_af(gene='KRAS', mutation='G12D')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69055e76",
   "metadata": {},
   "source": [
    "## Use case 2: Get all publications for a variant given the rsid\n",
    "- using rsid returned by the GDC /ssms endpt get all publications for a variant using litvar2 API\n",
    "- e.g. https://api.gdc.cancer.gov/ssms/edd1ae2c-3ca9-52bd-a124-b09ed304fcc2?fields=consequence.transcript.annotation.dbsnp_rs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "9f4c6a82",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_variant_pubs(gene, mutation):\n",
    "    pmids = {}\n",
    "    ssm_id = get_ssm_id(gene=gene, mutation=mutation)\n",
    "    ssms_endpt = 'https://api.gdc.cancer.gov/ssms/{}?fields=consequence.transcript.annotation.dbsnp_rs'.format(ssm_id)\n",
    "    output = requests.get(ssms_endpt).json()\n",
    "    dbsnp_rs_ids = []\n",
    "    for item in output['data']['consequence']:\n",
    "        dbsnp_rs_ids.append(item['transcript']['annotation']['dbsnp_rs'])\n",
    "    dbsnp_rs_ids = set(dbsnp_rs_ids)\n",
    "    for rsid in dbsnp_rs_ids:\n",
    "        variant_id = 'litvar@' + rsid + '%23%23'\n",
    "        endpt = 'https://www.ncbi.nlm.nih.gov/research/litvar2-api/variant/get/{}/publications'.format(variant_id)\n",
    "        output = requests.get(endpt).json()\n",
    "        pmids[rsid] = {'pmids': output['pmids'], 'pmids_count': output['pmids_count']}\n",
    "    # print output info\n",
    "    for k in pmids.keys():\n",
    "        print('number of pmids for {}: {}'.format(k, pmids[k]['pmids_count']))\n",
    "    return pmids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "2c9ede23",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of pmids for rs121913529: 13163\n"
     ]
    }
   ],
   "source": [
    "pmids = get_variant_pubs(gene='KRAS', mutation='G12D')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "746628cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rs121913529</th>\n",
       "      <td>28639239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs121913529</th>\n",
       "      <td>32604167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs121913529</th>\n",
       "      <td>29720585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs121913529</th>\n",
       "      <td>18645002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs121913529</th>\n",
       "      <td>32636940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs121913529</th>\n",
       "      <td>31817717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs121913529</th>\n",
       "      <td>31358966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs121913529</th>\n",
       "      <td>36634615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs121913529</th>\n",
       "      <td>25985019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs121913529</th>\n",
       "      <td>25788415</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13163 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                pmids\n",
       "rs121913529  28639239\n",
       "rs121913529  32604167\n",
       "rs121913529  29720585\n",
       "rs121913529  18645002\n",
       "rs121913529  32636940\n",
       "...               ...\n",
       "rs121913529  31817717\n",
       "rs121913529  31358966\n",
       "rs121913529  36634615\n",
       "rs121913529  25985019\n",
       "rs121913529  25788415\n",
       "\n",
       "[13163 rows x 1 columns]"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(pmids).T.explode('pmids').iloc[:, :1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca2429bb",
   "metadata": {},
   "source": [
    "### Use case 3: Query civic to get therapies for a variant\n",
    "- Consider gene and mutation entities, KRAS G12D\n",
    "- https://api.gdc.cancer.gov/ssms/edd1ae2c-3ca9-52bd-a124-b09ed304fcc2?fields=clinical_annotations.civic.variant_id\n",
    "- query civic API using these entities and format results in a df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "c3af78ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aartiv/.pyenv/versions/3.9.2/lib/python3.9/site-packages/gql/transport/aiohttp.py:92: UserWarning: WARNING: By default, AIOHTTPTransport does not verify ssl certificates. This will be fixed in the next major version. You can set ssl=True to force the ssl certificate verification or ssl=False to disable this warning\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "transport = AIOHTTPTransport(url=\"https://civicdb.org/api/graphql\")\n",
    "client = Client(transport=transport, fetch_schema_from_transport=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "15ecf1d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "async def get_therapies(input):\n",
    "    query = gql(\n",
    "        \"\"\"\n",
    "        query gene_variant_by_name($entrezSymbol: String!, $name: String!){\n",
    "            gene(entrezSymbol: $entrezSymbol) {\n",
    "                variants(name: $name) {\n",
    "                    nodes {\n",
    "                        name\n",
    "                        id\n",
    "                        link\n",
    "                        molecularProfiles {\n",
    "                            nodes {\n",
    "                                id\n",
    "                                name\n",
    "                                evidenceItems {\n",
    "                                    nodes {\n",
    "                                        id\n",
    "                                        status\n",
    "                                        phenotypes {\n",
    "                                            id\n",
    "                                            hpoId\n",
    "                                            name\n",
    "                                        }\n",
    "                                        description\n",
    "                                        therapies {\n",
    "                                            id\n",
    "                                            ncitId\n",
    "                                            name\n",
    "                                            therapyAliases\n",
    "                                        }\n",
    "                                        source {\n",
    "                                            ascoAbstractId\n",
    "                                            citationId\n",
    "                                            pmcId\n",
    "                                            sourceType\n",
    "                                            title\n",
    "                                        }\n",
    "                                        therapyInteractionType\n",
    "                                    }\n",
    "                                }\n",
    "                            }\n",
    "                        }\n",
    "                    }\n",
    "                }\n",
    "            }\n",
    "        }\n",
    "        \"\"\"\n",
    "    )\n",
    "    result = await client.execute_async(query, variable_values=input)\n",
    "    return result\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "dd4ed5ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "input = {'entrezSymbol': 'KRAS', 'name': 'G12D'}\n",
    "result = await get_therapies(input)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "57b1f359",
   "metadata": {},
   "outputs": [],
   "source": [
    "# format data for df\n",
    "df_data = {}\n",
    "df_data['id'] = []\n",
    "df_data['name'] = []\n",
    "df_data['therapies'] = []\n",
    "for item in result['gene']['variants']['nodes'][0]['molecularProfiles']['nodes']:\n",
    "    df_data['id'].append(item['id'])\n",
    "    df_data['name'].append(item['name'])\n",
    "    for evidence_items in item['evidenceItems']['nodes']:\n",
    "        therapy_list = [ therapy_details['name'] for therapy_details in evidence_items['therapies']]\n",
    "    df_data['therapies'].append(therapy_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "17b3b7a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>therapies</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>79</td>\n",
       "      <td>KRAS G12D</td>\n",
       "      <td>KRAS G12D Inhibitor HRS-4642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4571</td>\n",
       "      <td>KRAS G12D AND MET Splice Site (c.2888_3028del)</td>\n",
       "      <td>Crizotinib</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4632</td>\n",
       "      <td>KRAS G12D AND ERBB2 S310F</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4634</td>\n",
       "      <td>KRAS G12D AND ERBB2 S423R</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4635</td>\n",
       "      <td>ERBB2 R678Q AND KRAS G12D</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4638</td>\n",
       "      <td>KRAS G12D AND ERBB2 Q679L</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4640</td>\n",
       "      <td>KRAS G12D AND ERBB2 E717D</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>4641</td>\n",
       "      <td>ERBB2 L755S AND KRAS G12D</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>4642</td>\n",
       "      <td>ERBB2 V777L AND KRAS G12D</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>4643</td>\n",
       "      <td>ERBB2 V842I AND KRAS G12D</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>4762</td>\n",
       "      <td>KRAS G12D AND KRAS G12R AND CDKN2A Loss AND CD...</td>\n",
       "      <td>Palbociclib</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>4762</td>\n",
       "      <td>KRAS G12D AND KRAS G12R AND CDKN2A Loss AND CD...</td>\n",
       "      <td>Trametinib</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>4762</td>\n",
       "      <td>KRAS G12D AND KRAS G12R AND CDKN2A Loss AND CD...</td>\n",
       "      <td>Bevacizumab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>4763</td>\n",
       "      <td>KRAS G12D AND KRAS G12R AND CDKN2A Loss AND CD...</td>\n",
       "      <td>Palbociclib</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>4763</td>\n",
       "      <td>KRAS G12D AND KRAS G12R AND CDKN2A Loss AND CD...</td>\n",
       "      <td>Trametinib</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>4763</td>\n",
       "      <td>KRAS G12D AND KRAS G12R AND CDKN2A Loss AND CD...</td>\n",
       "      <td>Bevacizumab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>4768</td>\n",
       "      <td>KRAS G12D AND CDKN2A p16 R80*</td>\n",
       "      <td>Palbociclib</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>4768</td>\n",
       "      <td>KRAS G12D AND CDKN2A p16 R80*</td>\n",
       "      <td>Trametinib</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>4768</td>\n",
       "      <td>KRAS G12D AND CDKN2A p16 R80*</td>\n",
       "      <td>Afatinib</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>4775</td>\n",
       "      <td>KRAS G12D AND GNAS R201C AND NF1 D1976fs</td>\n",
       "      <td>Trametinib</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      id                                               name  \\\n",
       "0     79                                          KRAS G12D   \n",
       "1   4571     KRAS G12D AND MET Splice Site (c.2888_3028del)   \n",
       "2   4632                          KRAS G12D AND ERBB2 S310F   \n",
       "3   4634                          KRAS G12D AND ERBB2 S423R   \n",
       "4   4635                          ERBB2 R678Q AND KRAS G12D   \n",
       "5   4638                          KRAS G12D AND ERBB2 Q679L   \n",
       "6   4640                          KRAS G12D AND ERBB2 E717D   \n",
       "7   4641                          ERBB2 L755S AND KRAS G12D   \n",
       "8   4642                          ERBB2 V777L AND KRAS G12D   \n",
       "9   4643                          ERBB2 V842I AND KRAS G12D   \n",
       "10  4762  KRAS G12D AND KRAS G12R AND CDKN2A Loss AND CD...   \n",
       "10  4762  KRAS G12D AND KRAS G12R AND CDKN2A Loss AND CD...   \n",
       "10  4762  KRAS G12D AND KRAS G12R AND CDKN2A Loss AND CD...   \n",
       "11  4763  KRAS G12D AND KRAS G12R AND CDKN2A Loss AND CD...   \n",
       "11  4763  KRAS G12D AND KRAS G12R AND CDKN2A Loss AND CD...   \n",
       "11  4763  KRAS G12D AND KRAS G12R AND CDKN2A Loss AND CD...   \n",
       "12  4768                      KRAS G12D AND CDKN2A p16 R80*   \n",
       "12  4768                      KRAS G12D AND CDKN2A p16 R80*   \n",
       "12  4768                      KRAS G12D AND CDKN2A p16 R80*   \n",
       "13  4775           KRAS G12D AND GNAS R201C AND NF1 D1976fs   \n",
       "\n",
       "                       therapies  \n",
       "0   KRAS G12D Inhibitor HRS-4642  \n",
       "1                     Crizotinib  \n",
       "2                            NaN  \n",
       "3                            NaN  \n",
       "4                            NaN  \n",
       "5                            NaN  \n",
       "6                            NaN  \n",
       "7                            NaN  \n",
       "8                            NaN  \n",
       "9                            NaN  \n",
       "10                   Palbociclib  \n",
       "10                    Trametinib  \n",
       "10                   Bevacizumab  \n",
       "11                   Palbociclib  \n",
       "11                    Trametinib  \n",
       "11                   Bevacizumab  \n",
       "12                   Palbociclib  \n",
       "12                    Trametinib  \n",
       "12                      Afatinib  \n",
       "13                    Trametinib  "
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(df_data).explode('therapies')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
